### PR TITLE
Add a mesh modifier for adding sidesets using parsed condition 

### LINKED ISSUE #7930: ParsedSidesetMeshmodifier
### Description of the enhancement or error report
I would like to get some opinions if this capability should be added. It is similar to 

https://github.com/idaholab/moose/blob/devel/framework/src/meshmodifiers/ParsedSubdomainMeshModifier.C

but it adds Sidesets. In my case I want to add an interior side set between two subdomains,
but I would like to split this Sideset into multiple different Sidesets. I think there is currently
no other way than to actually split the subdomain and use `SideSetsBetweenSubdomains`. I don't want to do that because it would be messy and in certain instances in reactor physics the number of subdomains can be large reducing Moose performance. 

### Rationale for the enhancement or information for reproducing the error
I am not sure if this is generally useful or too particular in which case I can find another solution. 

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Question if this object is useful enough or just ballast. 

pinging @YaqiWang

### PR DESCRIPTION
**Pull request by snschune**
A mesh modifier that adds element's sides to a new sideset based on the following conditions:
1.  the centroid of the face satisfies the parsed condition
2. (optional) the owning element belongs to a listed subdomain
3. (optional) normal is to some tolerance equal to the provided normal

closes #7930

### CODE REVIEW

File: framework/include/meshmodifiers/ParsedAddSideset.h
Comment (dschwen) [inline framework/include/meshmodifiers/ParsedAddSideset.h]: look like this comment stops mid sentence

File: framework/include/meshmodifiers/ParsedAddSideset.h
Comment (dschwen) [inline framework/include/meshmodifiers/ParsedAddSideset.h]: That comment needs to be updated

Comment (dschwen) [inline framework/include/meshmodifiers/ParsedAddSideset.h]: That comment needs to be updated

File: framework/src/meshmodifiers/ParsedAddSideset.C
Comment (dschwen) [inline framework/src/meshmodifiers/ParsedAddSideset.C]: you could do the isParamSetByUser check here and give it a a default value of `Point()` for one less check below

**Review by permcody — CHANGES_REQUESTED** [review]
Please add this call to your `modify()` method:

```
_mesh_ptr->errorIfDistributedMesh("ParsedAddSideset");
```

This modifier won't work with ParallelMesh. You'll also need to restrict your new test so it won't run on the parallel mesh tests.

**Review by permcody — APPROVED** [review]

**Comment by snschune** [comment]
I need to fix the test before this is merged. The correct thing to do is to add

```
mesh_mode = REPLICATED
```

and I just added a cli.

**Comment by permcody** [comment]
I missed that! Thanks for self disclosing. I'll hold off on merging until this is in.

**Comment by friedmud** [comment]
This whole thing feels a bit specific to me.  I'm not sure it should actually be in MOOSE.  It sounds like a MeshModifier that should really just be in your app... or, at the most, in modules?

Someone else might have a different idea of how they would want to add sidesets based on parsed functions.  This one is a bit specific to your use-case

**Comment by WilkAndy** [comment]
Commenting on @friedmud 's comment that this MeshModifier is rather specific....  I'll probably use this particular MeshModifier, as I'm always mucking around adding sidesets to meshes (i often use VTK to build my meshes and it doesn't have good exodus support in this regard).  So, if you decide to move this contribution out of MOOSE, please put it in modules rather than in an App.

**Comment by snschune** [comment]
Hmmmm, it really doesn't fit into any module (since they separate by physics) except maybe misc. I'd be okay to have it in Yak or any other app I have access to. I have use for it in one particular application right now and that's it. I thought there might be a broader applicability to the mesh modifier the more I thought about it. @WilkAndy seems to confirm that this might be the case. In any case, I don't feel too strongly either way...

**Comment by permcody** [comment]
I suggest we keep it in the framework despite it being a little specific. If two completely different applications can use it, that's general enough.

**Comment by friedmud** [comment]
My vote was just that: a vote... if it's generally useful then go ahead and
put it in MOOSE.

It's hard to draw the line... but MOOSE will get HUGE if we don't curate it
somewhat...

**Pull request merged**
